sdk/keyvault/Azure.Security.KeyVault.Keys/src/KeysEventSource.cs

```diff
@@ -0,0 +1,96 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Diagnostics.Tracing;
+using System.Globalization;
+using System.Text;
+using Azure.Core.Diagnostics;
+
+namespace Azure.Security.KeyVault.Keys
+{
+    [EventSource(Name = EventSourceName)]
+    internal sealed class KeysEventSource : EventSource
+    {
+        internal const int AlgorithmNotSupportedEvent = 1;
+        internal const int KeyTypeNotSupportedEvent = 2;
+        internal const int PrivateKeyRequiredEvent = 3;
+        internal const int CryptographicExceptionEvent = 4;
+
+        private const string EventSourceName = "Azure-Security-KeyVault-Keys";
+
+        private KeysEventSource() : base(EventSourceName, EventSourceSettings.Default, AzureEventSourceListener.TraitName, AzureEventSourceListener.TraitValue) { }
+
+        public static KeysEventSource Singleton { get; } = new KeysEventSource();
+
+        [NonEvent]
+        public void AlgorithmNotSupported<T>(string operation, ref T algorithm) where T : struct
```

Comment:
is `ref T` overload required?

Reply:
Any reason to pass it by value? Requires an extra allocation, albeit small for which structs are intended to be passed here.

Reply:
It's the same stack size and it's simpler.